core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -198,14 +216,17 @@ class TransactionCoordinator(brokerId: Int,
           Left(Errors.CONCURRENT_TRANSACTIONS)
 
         case CompleteAbort | CompleteCommit | Empty =>
-          val transitMetadata = if (txnMetadata.isProducerEpochExhausted) {
-            val newProducerId = producerIdManager.generateProducerId()
-            txnMetadata.prepareProducerIdRotation(newProducerId, transactionTimeoutMs, time.milliseconds())
-          } else {
-            txnMetadata.prepareIncrementProducerEpoch(transactionTimeoutMs, time.milliseconds())
-          }
+          val transitMetadataResult =
+            if (txnMetadata.isProducerEpochExhausted &&
+                expectedProducerIdAndEpoch.forall(idAndEpoch => TransactionMetadata.isEpochExhausted(idAndEpoch.epoch))) {
```

Comment:
I think this is detecting the retry case. Another way to write it might be this:
```scala
expectedProducerIdAndEpoch.forall(_.epoch == txnMetadata.producerEpoch)
```